File: ruby/middleware.rb
require 'jwt'
require 'openssl'
require 'rack'

class WebhookSignatureMiddleware
  def initialize(app, secret, issuer, logger)
      @app = app
      @secret = secret
      @issuer = issuer
      @logger = logger
  end

  def call(env)
    request = Rack::Request.new(env)

    token = env['HTTP_X_WEBHOOK_SIGNATURE']
    unless token
      @logger.warn "❌ X-Webhook-Signature header not present"
      return [401, {}, ['Unauthorized']]
    end
    
    begin
      decoded_token = JWT.decode token, @secret, true, { algorithm: 'HS256' }
    rescue => err
      @logger.warn "❌ Could not decode JWT: #{err}"
      return [403, {}, ['Invalid JWT']]
    end
  
    claims = decoded_token[0]
  
    if claims['iss'] != @issuer
      @logger.warn "❌ Incorrect issuer: #{claims['iss']}"
      return [403, {}, ['Invalid JWT issuer']]
    end
    
    body = request.body.read
  
    expected_hash = OpenSSL::Digest::SHA256.digest(body).unpack1('H*')
  
    unless Rack::Utils.secure_compare(expected_hash, claims["sha256"])
      @logger.warn '❌ Body SHA256 does not match'
      @logger.warn "   JWT: #{claims['sha256']}"
      @logger.warn "Actual: #{expected_hash}"
      return [403, {}, ['Invalid JWT body hash']]
    end

    @logger.info '✅ X-Webhook-Signature Valid'

    request.body.rewind

    @app.call(env)
  end
end

File: python/middleware.py
import hashlib
import logging
import jwt
from werkzeug.wrappers import Request, Response

class WebhookSignatureMiddleware:
    def __init__(self, app, secret, issuer):
        self.app = app
        self.secret = secret
        self.issuer = issuer

    def __call__(self, environ, start_response):
        request = Request(environ)

        try:
            token = request.headers['X-Webhook-Signature']
            decoded_token = jwt.decode(token, self.secret, algorithms="HS256")
        except KeyError:
          logging.warn("❌ X-Webhook-Signature header not present")
          return Response(u'Unauthorized', mimetype= 'text/plain', status=401)(environ, start_response)
        except jwt.exceptions.InvalidSignatureError as err:
          logging.warn("❌ Could not decode JWT: %s", err)
          return Response(u'Invalid JWT', mimetype= 'text/plain', status=403)(environ, start_response)

        if decoded_token["iss"] != self.issuer:
          logging.warn("❌ Incorrect issuer: %s", decoded_token["iss"])
          return Response(u'Invalid JWT issuer', mimetype= 'text/plain', status=403)(environ, start_response)

        expected_hash = hashlib.sha256(request.data).hexdigest()

        if decoded_token["sha256"] != expected_hash:
          logging.warn('❌ Body SHA256 does not match')
          logging.warn("   JWT: %s", decoded_token["sha256"])
          logging.warn("Actual: %s", expected_hash)
          return Response(u'Invalid JWT body hash', mimetype= 'text/plain', status=403)(environ, start_response)

        logging.info('✅ X-Webhook-Signature Valid')
        return self.app(environ, start_response)
